test/unit/org/apache/cassandra/net/ConnectionTest.java

```diff
@@ -766,40 +767,84 @@ public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
+            // The total overly acquired amount divides the amount acquired in each step. Get the ceil value so not to miss the acquire that just exceeds.
+            long maxFailures = (long) Math.ceil((acquireCount * acquireStep * 2 - maxSendQueueCapacity) / acquireStep); // The result must be in the range of lone
             AtomicLong acquisitionFailures = new AtomicLong();
-            for (int i = 0; i < 100; i++)
-            {
-                executor.submit(() -> {
-                    for (int j = 0; j < 10000; j++)
-                    {
-                        if (!outbound.unsafeAcquireCapacity(acquireStep))
-                            acquisitionFailures.incrementAndGet();
-                    }
+            Runnable acquirer = () -> {
+                for (int j = 0; j < attempts; j++)
+                {
+                    if (!outbound.unsafeAcquireCapacity(acquireStep))
+                        acquisitionFailures.incrementAndGet();
+                }
+            };
+            Runnable releaser = () -> {
+                for (int j = 0; j < attempts; j++)
+                    outbound.unsafeReleaseCapacity(acquireStep);
+            };
 
-                });
-            }
+            // Start N acquirer and releaser to contend for capcaity
+            List<Runnable> invokeOrder = new ArrayList<>();
+            for (int i = 0 ; i < concurrency ; ++i)
+                invokeOrder.add(acquirer);
+            for (int i = 0 ; i < concurrency ; ++i)
+                invokeOrder.add(releaser);
+            // randomize their start order
+            randomize(invokeOrder);
 
-            for (int i = 0; i < 100; i++)
+            try
             {
-                executor.submit(() -> {
-                    for (int j = 0; j < 10000; j++)
-                        outbound.unsafeReleaseCapacity(acquireStep);
-                });
-            }
+                // Reserve enough capacity upfront to ensure the releaser threads cannot release all reserved capacity.
+                // i.e. the pendingBytes is always positive during the test.
+                Assert.assertTrue(outbound.unsafeAcquireCapacity(acquireCount, acquireCount * acquireStep));
+                ExecutorService executor = Executors.newFixedThreadPool(concurrency);
 
-            executor.shutdown();
-            executor.awaitTermination(10, TimeUnit.SECONDS);
+                invokeOrder.forEach(executor::submit);
 
-            // We can release more than we acquire, which certainly should not happen in
-            // real life, but since it's a test just for acquisition and release, it is fine
-            Assert.assertEquals(100 * 10000 * acquireStep - (acquisitionFailures.get() * acquireStep), outbound.pendingBytes());
+                executor.shutdown();
+                executor.awaitTermination(10, TimeUnit.SECONDS);
```

Comment:
your currency is 100, but if you run the test with 1 core is 10seconds enough time?  How stable is this on the LOWER circle ci configs?

Reply:
So far so good. I have been using the LOWER config, and not seeing an error from the await time. 
I do not expect 10 seconds to be not enough. It can be easily increased if this becomes a failure cause.